src/coprocessor/dag/rpn_expr/impl_compare.rs

```diff
@@ -54,6 +65,95 @@ impl<C: Comparer> Clone for RpnFnCompare<C> {
 
 // ======
 
+pub const fn compare_fn<C: Comparer>() -> RpnFn {
+    fn run<C: Comparer>(
+        rows: usize,
+        ctx: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue> {
+        ArgConstructor::new(1, ArgConstructor::new(0, CompareEvaluator::<C>::new()))
+            .eval(Null, rows, ctx, payload)
+    }
+    RpnFn {
+        name: "RpnFnCompare",
+        args_len: 2,
+        fn_ptr: run::<C>,
+    }
+}
+
+pub fn compare<C: Comparer>(lhs: &Option<C::T>, rhs: &Option<C::T>) -> Result<Option<i64>> {
+    C::compare(lhs, rhs)
+}
+
+pub struct CompareEvaluator<C: Comparer> {
+    _phantom: std::marker::PhantomData<C>,
+}
+
+impl<C: Comparer> CompareEvaluator<C> {
+    fn new() -> Self {
+        Self {
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<C: Comparer> Evaluator for CompareEvaluator<C> {
+    #[inline]
+    fn eval<D: ArgDef>(
+        self,
+        def: D,
+        rows: usize,
+        ctx: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue> {
+        CompareFn::<C>::eval(def, rows, ctx, payload)
+    }
+}
+
+trait CompareFn<C: Comparer> {
+    fn eval(
+        self,
+        rows: usize,
+        ctx: &mut EvalContext,
+        payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue>;
+}
+
+impl<C: Comparer, D: ArgDef> CompareFn<C> for D {
+    default fn eval(
+        self,
+        _rows: usize,
+        _ctx: &mut EvalContext,
+        _payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue> {
+        panic!("Cannot apply CompareFn on {:?}", self)
+    }
+}
+
+impl<'a, C, Arg0, Arg1> CompareFn<C> for Arg<Arg0, Arg<Arg1, Null>>
+where
+    C: Comparer,
+    Arg0: RpnFnArg<Type = &'a Option<C::T>>,
+    Arg1: RpnFnArg<Type = &'a Option<C::T>>,
+{
+    #[inline]
+    fn eval(
+        self,
+        rows: usize,
+        _ctx: &mut EvalContext,
+        _payload: RpnFnCallPayload<'_>,
+    ) -> Result<VectorValue> {
+        let arg = &self;
+        let mut result = Vec::with_capacity(rows);
+        for row in 0..rows {
+            let (arg0, arg) = arg.extract(row);
```

Comment:
My concern is whether or not the compiler can totally optimize line 149 ~ line 151. Could you provide a benchmark result?